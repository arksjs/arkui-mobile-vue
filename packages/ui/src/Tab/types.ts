import type {
  BadgeOption,
  BadgeProps as HandleBadgeOption
} from '../Badge/types'
import type { CSSProperties } from '../helpers'
import type { IconData } from '../Icon/types'

export interface OptionItem {
  label: string
  value: string | number
  icon?: string | IconData
  activeIcon?: string | IconData
  badge?: BadgeOption
  subLabel?: string
}

export type OptionList = (string | number | OptionItem)[]

export interface HandleOptionItem extends OptionItem {
  badge?: HandleBadgeOption
  iconLink?: string
  activeIconLink?: string
}

export type OnChange = (value: string | number, index: number) => void

export interface TabCommonProps {
  modelValue?: string | number
  options: OptionList
  color?: string
  activeColor?: string
  style?: CSSProperties
}

export interface TabCommonEmits {
  onChange?: OnChange
}

export interface TabProps extends TabCommonProps {
  scrollThreshold?: number
}

export type TabEmits = TabCommonEmits

export interface TabRef {
  switchTo: (value: string | number) => void
  switchToIndex: (index: number) => void
}

export type {
  OnChange as TabOnChange,
  OptionList as TabOptions,
  OptionItem as TabOption
}
