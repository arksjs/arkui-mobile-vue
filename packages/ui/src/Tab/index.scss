@import '../style/var.scss';

.ak-tab {
  --ak-color: #{$font-color};
  --ak-active-color: #{$primary-color};
  --ak-front-color: var(--ak-color);

  background-color: $white-color;
  height: 40px;
  position: relative;

  &_underline {
    position: absolute;
    bottom: 0;
    height: 2px;
    background-color: var(--ak-active-color);
    transition: all 0.2s;
  }

  &_list {
    padding: 0 4px;
    margin: 0;
    width: 100%;
    height: inherit;
    white-space: nowrap;
    box-sizing: border-box;
    position: relative;
    overflow-x: auto;
  }

  &_item {
    height: inherit;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-flow: column wrap;
    padding: 0 12px;
    position: relative;
    box-sizing: border-box;
    user-select: none;

    .ak-icon {
      --ak-icon-size: 20px;
      --ak-icon-color: var(--ak-front-color);

      margin-right: 4px;
    }

    &-inner {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
      position: relative;
      font-size: 16px;
      line-height: 22px;
      font-weight: 500;
      color: var(--ak-front-color);
    }

    &-text {
      display: block;
      word-break: break-all;
      white-space: pre-wrap;
    }

    &-sub-text {
      display: block;
      height: 17px;
      line-height: 17px;
      padding: 0 5px;
      border-radius: 2px;
      margin-top: 2px;
      font-size: 12px;
      color: var(--ak-front-color);
      opacity: 0.8;
    }

    &.active {
      --ak-front-color: var(--ak-active-color);

      .ak-tab_item {
        &-inner {
          font-weight: 700;
        }

        &-sub-text {
          background-color: var(--ak-front-color);
          color: #fff;
          opacity: 1;
        }
      }
    }
  }

  &.has--sub {
    height: 48px;

    .ak-tab {
      &_underline {
        display: none;
      }

      &_item {
        &::after {
          content: '';
          position: absolute;
          right: 0;
          top: 50%;
          height: 20px;
          margin-top: -10px;
        }

        &-inner {
          &::before {
            content: none;
          }
        }
      }
    }
  }

  &.no--scroll:not(.vertical) {
    .ak-tab {
      &_list {
        display: flex;
        padding: 0;
      }

      &_item {
        flex: 1;
        padding: 0;

        &-text {
          flex-grow: 1;
          display: -webkit-box;
          -webkit-line-clamp: 1;
          -webkit-box-orient: vertical;
          overflow: hidden;
        }
      }
    }
  }
}
