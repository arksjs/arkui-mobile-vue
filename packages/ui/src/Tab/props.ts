import type { PropType } from 'vue'
import { isNumber, isStringOrNumber, colorValidator, type PropsToEmits, isString } from '../helpers'
import type { OptionItem, OptionList, TabCommonEmits } from './types'

export const tabEmits: PropsToEmits<TabCommonEmits> = {
  'update:modelValue': value => isStringOrNumber(value),
  change: (value, index) => isStringOrNumber(value) && isNumber(index)
}

export const tabProps = {
  options: {
    type: Array as PropType<OptionList>,
    validator: (val: OptionList) => {
      if (Array.isArray(val)) {
        for (let i = 0; i < val.length; i++) {
          const option = val[i]

          if (isStringOrNumber(option)) {
            //
          } else if (option && isString(option.label) && isStringOrNumber(option.value)) {
            //
          } else {
            return false
          }
        }
      } else {
        return false
      }

      return true
    },
    required: true,
    default: () => [] as OptionItem[]
  },
  modelValue: {
    type: [Number, String]
  },
  color: {
    type: String,
    validator: colorValidator
  },
  activeColor: {
    type: String,
    validator: colorValidator
  }
}
