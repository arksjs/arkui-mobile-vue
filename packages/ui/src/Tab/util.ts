import type { CSSProperties } from '../helpers'

export const getStyles = (color?: string, activeColor?: string) => {
  const obj: CSSProperties = {}

  color && (obj['--ta-color'] = color)
  activeColor && (obj['--ta-active-color'] = activeColor)

  return obj
}

export const getClasses = (noScrolling: boolean, hasSub: boolean) => {
  return [
    'ta-tab',
    {
      'no--scrolling': noScrolling,
      'has--subtitle': hasSub
    }
  ]
}

export const getItemClasses = (index: number, activeIndex: number) => {
  return [
    'ta-tab_item',
    'ta-vertical-hairline',
    {
      active: index === activeIndex,
      'active-prev': index === activeIndex - 1,
      'active-next': index === activeIndex + 1
    }
  ]
}

export const TAB_HEIGHT = 40
