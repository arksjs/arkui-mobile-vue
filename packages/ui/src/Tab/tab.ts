import type { PropType } from 'vue'
import { isNumber, isStringNumberMix } from '../helpers/util'
import type { OptionItem, OptionList, TabCommonEmits } from './types'
import { colorValidator } from '../helpers/validator'
import type { PropsToEmits } from '../helpers/types'

export const tabEmits: PropsToEmits<TabCommonEmits> = {
  'update:activeValue': value => isStringNumberMix(value),
  change: (value, index) => isStringNumberMix(value) && isNumber(index)
}

export const tabProps = {
  options: {
    type: Array as PropType<OptionList>,
    validator: (val: OptionList) => {
      if (Array.isArray(val)) {
        for (let i = 0; i < val.length; i++) {
          const option = val[i]

          if (typeof option === 'string' || typeof option === 'number') {
            //
          } else if (
            option &&
            typeof option.label === 'string' &&
            isStringNumberMix(option.value)
          ) {
            //
          } else {
            return false
          }
        }
      } else {
        return false
      }

      return true
    },
    required: true,
    default: () => [] as OptionItem[]
  },
  activeValue: {
    type: [Number, String] as PropType<number | string>
  },
  color: {
    type: String,
    validator: colorValidator
  },
  activeColor: {
    type: String,
    validator: colorValidator
  }
}
