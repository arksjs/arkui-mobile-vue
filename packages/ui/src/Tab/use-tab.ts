import {
  getCurrentInstance,
  ref,
  watch,
  computed,
  nextTick,
  shallowRef
} from 'vue'
import type { SetupContext, ExtractPropTypes } from 'vue'
import { isNumber, isObject, isStringNumberMix, isURL } from '../helpers/util'
import { handleBadge } from '../Badge/util'
import type { OptionItem, HandleOptionItem } from './types'
import type { tabEmits, tabProps } from './tab'
import { getStyles } from './util'
import { useFrameTask } from '../hooks/use-frame-task'
import { useException } from '../hooks/use-exception'
import { useOnce } from '@/hooks/use-once'

interface UseOptions {
  tabName: string
}

export function useTab(
  props: ExtractPropTypes<typeof tabProps>,
  { emit, expose }: SetupContext<typeof tabEmits>,
  { tabName }: UseOptions
) {
  const { printNotInOptionsError } = useException()
  const instance = getCurrentInstance()
  const listEl = shallowRef<HTMLElement | null>(null)
  const underlineEl = shallowRef<HTMLElement | null>(null)
  const options2 = ref<HandleOptionItem[]>([])
  const activeIndex = ref(-1)
  const hasSub = ref(false)

  const { frameStart } = useFrameTask()

  let activeValue = props.modelValue ?? ''

  function updateOptions() {
    const options: HandleOptionItem[] = []

    let hasActive = false
    hasSub.value = false

    if (Array.isArray(props.options)) {
      props.options.forEach((item, index) => {
        let option: HandleOptionItem | null = null

        if (isNumber(item)) {
          option = {
            label: item.toString(),
            value: item as number
          }
        } else if (typeof item === 'string') {
          option = {
            label: item,
            value: item
          }
        } else if (isObject(item)) {
          item = item as OptionItem

          if (isStringNumberMix(item.value)) {
            option = {
              label:
                typeof item.label === 'string'
                  ? item.label
                  : item.value.toString(),
              subLabel: typeof item.subLabel === 'string' ? item.subLabel : '',
              value: item.value,
              icon: null
            }

            if (item.icon) {
              if (isURL(item.icon)) {
                option.iconLink = item.icon as string
                option.activeIconLink = isURL(item.activeIcon)
                  ? (item.activeIcon as string)
                  : option.iconLink
              } else {
                option.icon = item.icon
                option.activeIcon = item.activeIcon ?? option.icon
              }
            }

            if (option.subLabel) {
              hasSub.value = true
            }

            option.badge = handleBadge(item.badge)
          }
        }

        if (option) {
          if (!hasActive && option.value === activeValue) {
            // 找到原来的value， activeValue 和 activeIndex 都正确
            activeIndex.value = index
            hasActive = true
          }

          options.push(option)
        }
      })
    }

    options2.value = options

    if (!hasActive) {
      if (options[0]) {
        // 设置为第一个
        activeIndex.value = 0
        activeValue = options[0].value
      } else {
        activeIndex.value = -1
        activeValue = ''
      }
      // 首次属于prop传进来的，不emit回去
      instance?.isMounted && emitChange()
    }

    updatePos()
  }

  function _switchTo(value: string | number, isProp = false) {
    if (value === activeValue) {
      return
    }

    if (!updateActive(value)) {
      printNotInOptionsError('index', isProp)
    } else if (!isProp) {
      // 设置modelValue不调用onChange
      emitChange()
    }
  }

  function switchToIndex(index: number) {
    if (index === activeIndex.value) {
      return
    }

    if (options2.value[index]) {
      onChange(options2.value[index].value)
    } else {
      printNotInOptionsError('index')
    }
  }

  function updateActive(value: string | number) {
    if (value === activeValue) {
      return true
    }

    let hasValue = false

    options2.value.forEach((option, index) => {
      if (option.value === value) {
        activeValue = option.value
        activeIndex.value = index
        hasValue = true
      }
    })

    hasValue && instance?.isMounted && updatePos()

    return hasValue
  }

  function onChange(value: string | number) {
    if (value === activeValue) {
      return
    }

    updateActive(value)
    emitChange()
  }

  function emitChange() {
    emit('update:modelValue', activeValue)
    emit('change', activeValue, activeIndex.value)
  }

  function updatePos() {
    nextTick(() => {
      if (tabName === 'TabBar') {
        return
      }

      if (!listEl.value) {
        return
      }
      const $list = listEl.value
      const $activeItem = $list.children[activeIndex.value] as HTMLElement
      if (!$activeItem) {
        if (tabName === 'Tab') {
          updateUnderline()
        }
        return
      }

      const vertical = tabName === 'SideTab'
      const sizeKey = !vertical ? 'Width' : 'Height'
      const directionKey = !vertical ? 'Left' : 'Top'
      const offsetSizeKey = ('offset' + sizeKey) as 'offsetWidth'
      const scrollSizeKey = ('scroll' + sizeKey) as 'scrollWidth'
      const offsetDirectionKey = ('offset' + directionKey) as 'offsetLeft'
      const scrollDirectionKey = ('scroll' + directionKey) as 'scrollLeft'

      const w = $list[offsetSizeKey]
      const ofs = $activeItem[offsetDirectionKey]
      const from = $list[scrollDirectionKey]
      const to =
        $activeItem[offsetSizeKey] > w
          ? ofs
          : Math.max(
              Math.min(
                ofs - (w - $activeItem[offsetSizeKey]) / 2,
                $list[scrollSizeKey] - w
              ),
              0
            )

      // console.log(from, to, w, ofs, ofs - to)

      frameStart({
        from,
        to,
        duration: 200,
        progress({ current }) {
          $list[scrollDirectionKey] = current
        },
        complete({ current }) {
          $list[scrollDirectionKey] = current
        }
      })

      if (tabName === 'Tab') {
        updateUnderline()
      }
    })
  }

  const updateUnderlineOnce = useOnce(50)

  function updateUnderline() {
    updateUnderlineOnce(() => {
      const $activeItem = listEl.value?.querySelector('.active') as HTMLElement

      let x = 0
      let w = 0

      if ($activeItem) {
        const $inner = $activeItem.firstElementChild as HTMLElement

        x =
          $activeItem.offsetLeft -
          (listEl.value?.scrollLeft || 0) +
          ($activeItem.offsetWidth - $inner.offsetWidth) / 2
        w = $inner.offsetWidth
      }

      if (underlineEl.value) {
        underlineEl.value.style.width = w + 'px'
        underlineEl.value.style.transform = `translate3d(${x}px, 0, 0)`
      }
    })
  }

  watch(
    () => props.modelValue,
    val => val != null && _switchTo(val, true)
  )

  watch(() => props.options, updateOptions, {
    deep: true,
    immediate: true
  })

  const styles = computed(() => getStyles(props.color, props.activeColor))

  const switchTo = (value: string | number) => _switchTo(value, false)

  expose({
    switchTo,
    switchToIndex
  })

  return {
    listEl,
    underlineEl,
    activeIndex,
    hasSub,
    options2,
    onChange,
    styles,
    updateUnderline,

    switchTo,
    switchToIndex
  }
}
