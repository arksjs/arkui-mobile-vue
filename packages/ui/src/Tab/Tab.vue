<template>
  <div :class="classes">
    <ul class="fx-tab_list" ref="listEl">
      <li
        :class="getItemClasses(index, activeIndex)"
        v-for="(item, index) in options2"
        :key="item.value"
        @click="onChange(item.value)"
      >
        <Badge class="fx-tab_item-inner" v-bind="item.badge">
          <Icon
            v-if="item.icon"
            :icon="index === activeIndex ? item.activeIcon : item.icon"
          />
          <span class="fx-tab_item-text">{{ item.label }}</span>
        </Badge>
        <span class="fx-tab_item-sub-text" v-if="hasSub">{{
          item.subLabel
        }}</span>
      </li>
    </ul>
    <span class="fx-tab_underline" ref="underlineEl"></span>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue'
import { Icon } from '../Icon'
import { Badge } from '../Badge'
import { tabEmits, tabProps } from '../Tab/tab'
import { useTab } from '../Tab/use-tab'
import { getClasses, getItemClasses } from './util'

export default defineComponent({
  name: 'fx-tab',
  components: { Icon, Badge },
  props: {
    ...tabProps,
    scrollThreshold: {
      type: Number,
      default: 4
    }
  },
  emits: { ...tabEmits },
  setup(props, ctx) {
    const tab = useTab(props, ctx, { tabName: 'Tab' })

    const classes = computed(() =>
      getClasses(props.scrollThreshold, tab.options2.value, tab.hasSub.value)
    )

    return {
      classes,
      getItemClasses,
      ...tab
    }
  }
})
</script>
