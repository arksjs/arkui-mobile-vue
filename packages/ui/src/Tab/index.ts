import { withInstall } from '../helpers/with-install'
import Tab from './Tab.vue'

export { Tab }
export const install = withInstall(Tab)
export default Tab
