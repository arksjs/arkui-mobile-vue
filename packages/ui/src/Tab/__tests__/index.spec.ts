import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Tab } from '@/Tab'
import { mixTabList, shortTabList, subTabList } from './data'
import { nextTick } from 'vue'

describe('Tab', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Tab, {
      props: {
        options: []
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('snapshot mixTabList', () => {
    const wrapper = mount(Tab, {
      props: {
        options: mixTabList
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('snapshot mixTabList', () => {
    const wrapper = shallowMount(Tab, {
      props: {
        options: mixTabList
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('snapshot subTabList', () => {
    const wrapper = mount(Tab, {
      props: {
        options: subTabList
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('switchTo & switchToIndex', async () => {
    const wrapper = mount(Tab, {
      props: {
        options: shortTabList
      }
    })

    wrapper.findComponent(Tab).vm.switchToIndex(1)
    await nextTick()
    expect(wrapper.findAll('.ak-tab_item')[1].classes()).toContain('active')

    wrapper.findComponent(Tab).vm.switchTo(4)
    await nextTick()
    expect(wrapper.findAll('.ak-tab_item')[3].classes()).toContain('active')
  })
})
