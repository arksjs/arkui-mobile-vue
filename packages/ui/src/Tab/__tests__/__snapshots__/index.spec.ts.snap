// Vitest Snapshot v1

exports[`Tab > snapshot 1`] = `
"<div class=\\"fx-tab no--scroll\\">
  <ul class=\\"fx-tab_list\\"></ul><span class=\\"fx-tab_underline\\"></span>
</div>"
`;

exports[`Tab > snapshot mixTabList 1`] = `
"<div class=\\"fx-tab no--scroll\\">
  <ul class=\\"fx-tab_list\\">
    <li class=\\"fx-tab_item fx-vertical-hairline active\\">
      <div class=\\"fx-badge fx-tab_item-inner\\"><svg class=\\"fx-icon\\">
          <use xlink:href=\\"#CheckSquareFilled\\"></use>
        </svg><span class=\\"fx-tab_item-text\\">买过</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline active-next\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">秒杀</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(1); right: 0px; top: 0px;\\">惠</div>
      </div>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">天天半价工作餐</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">套餐</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(1); right: 0px; top: 0px;\\">1</div>
      </div>
      <!--v-if-->
    </li>
  </ul><span class=\\"fx-tab_underline\\"></span>
</div>"
`;

exports[`Tab > snapshot mixTabList 2`] = `
"<div class=\\"fx-tab no--scroll\\">
  <ul class=\\"fx-tab_list\\">
    <li class=\\"fx-tab_item fx-vertical-hairline active\\">
      <badge-stub content=\\"0\\" maxcount=\\"99\\" dot=\\"false\\" showzero=\\"false\\" animated=\\"false\\" offset=\\"0,0\\" class=\\"fx-tab_item-inner\\"></badge-stub>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline active-next\\">
      <badge-stub content=\\"惠\\" maxcount=\\"99\\" dot=\\"false\\" showzero=\\"false\\" animated=\\"false\\" offset=\\"0,0\\" class=\\"fx-tab_item-inner\\"></badge-stub>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <badge-stub content=\\"0\\" maxcount=\\"99\\" dot=\\"false\\" showzero=\\"false\\" animated=\\"false\\" offset=\\"0,0\\" class=\\"fx-tab_item-inner\\"></badge-stub>
      <!--v-if-->
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <badge-stub content=\\"1\\" maxcount=\\"99\\" dot=\\"false\\" showzero=\\"false\\" animated=\\"false\\" offset=\\"0,0\\" class=\\"fx-tab_item-inner\\"></badge-stub>
      <!--v-if-->
    </li>
  </ul><span class=\\"fx-tab_underline\\"></span>
</div>"
`;

exports[`Tab > snapshot subTabList 1`] = `
"<div class=\\"fx-tab no--scroll has--sub\\">
  <ul class=\\"fx-tab_list\\">
    <li class=\\"fx-tab_item fx-vertical-hairline active\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">精选</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div><span class=\\"fx-tab_item-sub-text\\">为你推荐</span>
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline active-next\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">折扣</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div><span class=\\"fx-tab_item-sub-text\\">今日折扣</span>
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">好物</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div><span class=\\"fx-tab_item-sub-text\\">好物必备</span>
    </li>
    <li class=\\"fx-tab_item fx-vertical-hairline\\">
      <div class=\\"fx-badge fx-tab_item-inner\\">
        <!--v-if--><span class=\\"fx-tab_item-text\\">新品</span>
        <div class=\\"fx-badge_badge\\" style=\\"transform: translate3d(50%, -50%, 0px) scale(0); right: 0px; top: 0px;\\">0</div>
      </div><span class=\\"fx-tab_item-sub-text\\">新品速递</span>
    </li>
  </ul><span class=\\"fx-tab_underline\\"></span>
</div>"
`;
